### ISSUE TITLE
 Types for complex.__new__ should be any numeric type

### ISSUE THREAD

**Issue by smlerman** [python/typeshed#8440]
The types for the arguments to `complex.__new__` should be any numeric type, not just float.

```
c1 = complex(1, 2)
c2 = complex(1, c1)
```

This code is valid, but the second line reports an error that complex is incompatible with float.

https://github.com/python/typeshed/blob/f550c24886d5daa84b70fe2960b93fa05ea2eb0b/stdlib/builtins.pyi#L368

**Comment by AlexWaygood** [comment]
The [docs for `complex`](https://docs.python.org/3/library/functions.html#complex) state:

>  Each argument may be any numeric type (including complex).

So this seems correct. PR welcome!

**Comment by smlerman** [comment]
I tried changing `float` to `numbers.Number`, but I got errors that `int`, `float`, and `complex` are all incompatible. Is that expected? If so, changing it to `float | complex` seems to work correctly.

**Comment by AlexWaygood** [comment]
> I tried changing `float` to `numbers.Number`, but I got errors that `int`, `float`, and `complex` are all incompatible. Is that expected? If so, changing it to `float | complex` seems to work correctly.

Yeah, that's expected (see the resources I laid out [here](https://github.com/python/typeshed/issues/8348#issuecomment-1190658389) for why `numbers.Number` isn't much use when it comes to static typing.

For the first argument, the docs state:

> For a general Python object x, `complex(x)` delegates to `x.__complex__()`. If `__complex__()` is not defined then it falls back to `__float__()`. If `__float__()` is not defined then it falls back to `__index__()`.

That means that the first argument should be `complex | SupportsComplex | SupportsFloat | SupportsIndex`.

For the second argument, it seems `complex` or anything with a `__float__` method will do, so the argument should be `complex | SupportsFloat`:

```pycon
>>> class Foo:
...   def __index__(self):
...     return 5
...     
>>> complex(1, Foo())
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: complex() second argument must be a number, not 'Foo'
>>> class Foo:
...   def __complex__(self):
...     return 5
...     
>>> complex(1, Foo())
Traceback (most recent call last):
  File "<string>", line 1, in <module>
TypeError: complex() second argument must be a number, not 'Foo'
>>> class Foo:
...   def __float__(self):
...     return 5
...     
>>> complex(1, Foo())
(1+5j)
```

**Comment by AlexWaygood** [comment]
`SupportsComplex` and `SupportsFloat` are both importable from `typing`; `SupportsIndex` is importable from `typing_extensions`.

**Comment by AlexWaygood** [comment]
> For the first argument, the [docs](https://docs.python.org/3/library/functions.html#complex) state:
> 
> > For a general Python object x, `complex(x)` delegates to `x.__complex__()`. If `__complex__()` is not defined then it falls back to `__float__()`. If `__float__()` is not defined then it falls back to `__index__()`.
> 
> That means that the first argument should be `complex | SupportsComplex | SupportsFloat | SupportsIndex`.

Actually, looks like that's only the case for when you only provide a single argument to `complex()`. If you provide two arguments, then the first argument should also be `complex | SupportsFloat`.

**Comment by AlexWaygood** [comment]
This is more complicated than I realised; I can prepare the PR if you like :)

**Comment by smlerman** [comment]
I was going to ask what Python version you were using, because I was getting different results with your code, then I saw your comment about using 3.7.

There's already a one-parameter overload declared, which has `str | SupportsComplex | SupportsIndex | complex` (should maybe add `SupportsFloat`; I can test it out), so I think I'd only have to fix the two-parameter overload.

**Comment by smlerman** [comment]
complex(1, FooSupportsIndex()) only works in 3.8+. Is it worth putting in a version check with different declarations for <3.8 and >=3.8?

**Comment by AlexWaygood** [comment]
> complex(1, FooSupportsIndex()) only works in 3.8+. Is it worth putting in a version check with different declarations for <3.8 and >=3.8?

Yes, absolutely!

**Comment by smlerman** [comment]
Attached a script for checking everything

[python_typeshed_8440.py.txt](https://github.com/python/typeshed/files/9222755/python_typeshed_8440.py.txt)

**Comment by smlerman** [comment]
For some bizarre reason, Python gives `TypeError: complex() second argument must be a number, not 'FooComplex'` for a `SupportsComplex` object, even though a `complex` works fine. This is on all versions up to 3.10.

**Comment by AlexWaygood** [comment]
> For some bizarre reason, Python gives `TypeError: complex() second argument must be a number, not 'FooComplex'` for a `SupportsComplex` object, even though a `complex` works fine. This is on all versions up to 3.10.

Some aspects of Python's data model are a little rough around the edges :)

**Comment by smlerman** [comment]
I assume the declaration here should match Python's actual behavior, and not what it probably should be?

**Comment by AlexWaygood** [comment]
> I assume the declaration here should match Python's actual behavior, and not what it probably should be?

Yes, absolutely. "What the behaviour should be" is the  question you ask over at the CPython repo, not at typeshed :)

**Comment by smlerman** [comment]
Take a look at https://github.com/smlerman/python-typeshed/commit/bc7465901fa18153dee2ac9169ae93c5514d796f and let me know if that looks right.

**Comment by AlexWaygood** [comment]
> Take a look at [smlerman@bc74659](https://github.com/smlerman/python-typeshed/commit/bc7465901fa18153dee2ac9169ae93c5514d796f) and let me know if that looks right.

That looks great! The only thing to change is that `SupportsFloat` should be added to the union in both the `>= (3, 8)` branch and the `< (3, 8)` branch.

For the two-argument overload, it is _utterly_ bizarre that `complex` accepts `SupportsComplex` _for the first argument, but not the second_, but that does indeed seem to be the case!

**Comment by JelleZijlstra** [comment]
The C code (for 3.12) is here: https://github.com/python/cpython/blob/cd26595232ac1b5061460d5949d5204c90287c1c/Objects/complexobject.c#L898. It indeed uses `__complex__` only for the real argument, not the imaginary one.

**Comment by smlerman** [comment]
Just added https://github.com/smlerman/python-typeshed/commit/070a113b5cc0d4d72562834097f6961fadaed00c

**Comment by AlexWaygood** [comment]
Looks great! Please feel free to submit a PR with the fix :)

**Comment by AlexWaygood** [comment]
Fixed by #8473. Thanks @smlerman!
